[WIP] Ensure using no-double-extend does not modify an object's own keys.

rwjblue opened the pull request:
Given:

```js
owner.register('some:name', Ember.Object.extend());
let instance = owner.lookup('some:name');

assert.deepEqual(Object.keys(instance), []);
```

This assertion passes in Ember 2.12 and older, but not Ember 2.13.0 beta series. The cause of this regression was disabling `no-double-extend` feature and now some properties like `_debugContainerKey` and `NAME_KEY` are being set on new instances.

Commit:
Move `isObject` to `ember-utils`.

Commit:
Expose mechanism for instances to access their FactoryManager.

Commit:
Remove `Ember.get` usage from container tests.

Commit:
Move away from `NAME_KEY` for building `toString`.

Instead use the instances FactoryManager to access the toString value.

krisselden commented on packages/container/lib/container.js:
```diff
@@ -618,6 +647,18 @@ class FactoryManager {
+    }
+  }
+
+  factoryToString() {
```
why not just make this `toString() {}`

krisselden commented on packages/ember-runtime/lib/system/core_object.js:
```diff
@@ -540,7 +541,9 @@ CoreObject.PrototypeMixin = Mixin.create({
     let extension = hasToStringExtension ? `:${this.toStringExtension()}` : '';
-    let ret = `<${this[NAME_KEY] || this.constructor.toString()}:${guidFor(this)}${extension}>`;
+    let factoryManager = peekFactoryManager(Object.getPrototypeOf(this));
+    let factoryToString = this[NAME_KEY] || (factoryManager && factoryManager.factoryToString());
```
if `factoryToString()` was just toString() you wouldn't need to call it you can just let the string template call it.

krisselden commented on packages/container/lib/container.js:
```diff
@@ -618,6 +647,18 @@ class FactoryManager {
     this.injections = undefined;
+
+    if (isObject(factory)) {
+      setFactoryManager(factory, this);
```
factory here doesn't not have a 1 to 1 with FactoryManager, FactoryManager has only one factory but a factory can have multiple managers.

rwjblue replied:
Yeah, this is a fairly major issue for us. I'll ponder it a bit...

Commit:
Set the factory manager per instance created.

The prior implementation stored the factory manager on the factory itself.
Unfortunately, this means that a single factory can only have _one_
FactoryManager which breaks the scenario of rexporting the same factory
from multiple locations. In the reexport scenario, each reexport should
have its own `fullName`, injections, etc.

Commit:
Move `_debugContainerKey` out of injections.

For paths that go through `container.lookupFactory` (including
`owner._lookupFactory`) still add `_debugContainerKey` via `.extend`
(normal injections) and `.reopenClass` (factory injections).

For paths that go through `factoryFor` the `_debugContainerKey` is not
set in the `injections` args, and is only available through
`peekFactoryManager`. A helpful getter is added to `Ember.Object` so that
`this._debugContainerKey` continues to function properly.

The tests use a base factory that does not define a `_debugContainerKey`
getter (we cannot rely on `Ember.Object` in `container` package). In tests
that were confirming `_debugContainerKey` is present, I have switched them
to use `peekFactoryManager`.
